bip-0000.mediawiki

```diff
@@ -0,0 +1,335 @@
+
+Alice performs the tweak with the sum of her input private keys in the following manner:
+
+* Let ''outpoints_hash = hash(txid<sub>0</sub> || vout<sub>0</sub> || … txid<sub>n</sub> || vout<sub>n</sub>)''
+* Let ''a = a<sub>0</sub> + a<sub>1</sub> … + a<sub>n</sub>''
+* Let ''P<sub>0</sub> = hash(outpoints_hash·a·B || 0)·G + B''
+
+''' Spend and Scan Key '''
+
+Since Bob needs his private key ''b'' to check for incoming payments, this requires ''b'' to be exposed to an online device. To minimize the risks involved, Bob can instead publish an address of the form ''(B<sub>scan</sub>, B<sub>spend</sub>)''. This allows Bob to keep ''b<sub>spend</sub>'' in offline cold storage and perform the scanning with the public key ''B<sub>spend</sub>'' and private key ''b<sub>scan</sub>''. Alice performs the tweak using both of Bob's public keys in the following manner:
+
+* Let ''P<sub>0</sub> = hash(outpoints_hash·a·B<sub>scan</sub> || 0)·G + B<sub>spend</sub>''
+
+Bob detects this payment by calculating ''P<sub>0</sub> = hash(outpoints_hash·b<sub>scan</sub>·A)·G + B<sub>spend</sub>'' with his online device and can spend from his cold storage signing device using ''(hash(outpoints_hash·b<sub>scan</sub>·A) + b<sub>spend</sub>) mod p'' as the private key.
+
+''' Labels '''
+
+For a single silent payment address of the form ''(B<sub>scan</sub>, B<sub>spend</sub>)'', Bob may wish to differentiate incoming payments. Naively, Bob could publish multiple silent payment addresses, but this would require him to scan for each one, which becomes prohibitively expensive. Instead, Bob can label his spend public key ''B<sub>spend</sub>'' with an integer ''m'' in the following way:
+
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + m·G''
+* Publish ''(B<sub>scan</sub>, B<sub>0</sub>)'', ''(B<sub>scan</sub>, B<sub>1</sub>) …''
+
+Alice performs the tweak as before using one of the published ''(B<sub>scan</sub>, B<sub>m</sub>)'' pairs. Bob detects the labeled payment in the following manner:
+
+* Let ''P<sub>0</sub> = hash(outpoints_hash·b<sub>scan</sub>·A || 0)·G + B<sub>spend</sub>''
+* Subtract ''P<sub>0</sub>'' from each of the transaction outputs and check if the remainder matches any of the labels (''1·G, 2·G ..'') that the wallet has previously used
+
+It is important to note that an outside observer can easily deduce that each published ''(B<sub>scan</sub>, B<sub>m</sub>)'' pair is owned by the same entity as each published address will have ''B<sub>scan</sub>'' in common. As such, labels are not meant as a way for Bob to manage separate identities, but rather a way for Bob to determine the source of an incoming payment.
```

Comment:
This feels unnecessarily foot-shooty. Since `m=[1,2,3,...]` is arbitrary and only really important to Bob, why can't he simply do `m=[b<sub>scan</sub>*1, b<sub>scan</sub>*2, b<sub>scan</sub>*3,...]` instead? No additional information is required, and Bob's check would check for the above list, rather than the known-to-everybod `int*G` list.

Reply:
Hey @kallewoof, thanks for the review! With multiple `Bscans`, Bob now has to do the ECDH step multiple times for the same transaction to determine if there is an output for him. For someone with sufficiently powerful hardware, this might not be an issue, but it certainly wouldn't be possible on mobile or even your "average" node runner using commodity hardware.

Imagine the use case where Bob is an exchange and wants to have a unique SP address for each user: if he hands out a `Bscan*int` per user, he now has to do an EC mult per user, per transaction. With the labels approach we describe, he would only have to do one EC mult per transaction, subtract that from the outputs of the transaction and see if any of the remainders (`int*G + Bspend` ) are in his database with a simple lookup. This scales nicely to millions of users.

> rather than the known-to-everybod int*G list

I'm not sure what you mean by known-to-everybody. If Bob posts two SP addresses, `Bscan, Bspend + 1*G` and `Bscan, Bspend + 2*G`, an outside observer can see two SP addresses with `Bscan` in common and infer that they are owned by the same entity, but they don't know `Bspend` and can't determine which `int*G` tweak was used, or even if an `int*G` tweak was used.

Reply:
> Hey @kallewoof, thanks for the review! With multiple `Bscans`, Bob now has to do the ECDH step multiple times for the same transaction to determine if there is an output for him. For someone with sufficiently powerful hardware, this might not be an issue, but it certainly wouldn't be possible on mobile or even your "average" node runner using commodity hardware.

Is powmod really that expensive? I thought it was pretty instant/cheap.

> Imagine the use case where Bob is an exchange and wants to have a unique SP address for each user: if he hands out a `Bscan*int` per user, he now has to do an EC mult per user, per transaction. With the labels approach we describe, he would only have to do one EC mult per transaction, subtract that from the outputs of the transaction and see if any of the remainders (`int*G + Bspend` ) are in his database with a simple lookup. This scales nicely to millions of users.

If we're talking millions of users, then perhaps it becomes prohibitively expensive, but otoh an exchange won't be run off of a mobile device, so maybe they can handle that processing power bump. Perhaps I'm underestimating the complexity-bump here.

> I'm not sure what you mean by known-to-everybody. If Bob posts two SP addresses, `Bscan, Bspend + 1*G` and `Bscan, Bspend + 2*G`, an outside observer can see two SP addresses with `Bscan` in common and infer that they are owned by the same entity, but they don't know `Bspend` and can't determine which `int*G` tweak was used, or even if an `int*G` tweak was used.

If I see a pubkey X, and I then see a pubkey X + (known to everyone) n*G where n is a tiny predictable number, then I can indeed determine which `int*G` tweak was used, no? It would require me to track a lot of data, so perhaps it's no biggie, but it definitely isn't impossible.

Reply:
> Is powmod really that expensive? I thought it was pretty instant/cheap

It's more that EC addition is cheaper. The other significant advantage in my mind is that if we apply the additive tweak to `Bspend`, we only have to do this once and can cache it. If we use the `Bscan*int` approach, we have to redo the calculation for each `int` for each new transaction (i.e `Bscan*int*InputPubKeys`)

> so maybe they can handle that processing power bump

Perhaps worth mentioning there is nothing stopping an exchange (or anyone with sufficient processing power) from doing it this way and they would still be completely compatible with the SP protocol! The sender doesn't care (or even know) how the receiver is tweaking their keys. Its not clear to me why the exchange would prefer this over the additive method, but they could do it. It would be a custom implementation, which would make recovering from backup into different wallet software an issue, but that's more a concern for individual users, imo.

> If I see a pubkey X, and I then see a pubkey X + (known to everyone) n*G

Right, my point was you've never seen pubkey X (`Bspend`, in this example), and you wouldn't know it was the untweaked `Bspend` even if you saw it. All you see are three Pubkeys. 

I'd also add it doesn't hurt the receiver's privacy or security at all if you were to learn which `int*G` they used. It's just a nice benefit that until you've seen multiple addresses with `Bscan` in common, you can't know whether or not any particular SP addresses is using a label tweak or not.

Reply:
I think I got caught on the initial `"easily deduce that each published ''(B<sub>scan</sub>, B<sub>m</sub>)'' pair is owned by the same entity"` part. But if I'm not mistaken, the Bspend pubkeys are not used directly, but they are tweaked in payments, so even knowing them I won't know when a payment is made using them, unless I hold the corresponding privkey.

Thanks, resolving.

Reply:
Thanks for the comments @kallewoof.

>If I see a pubkey X, and I then see a pubkey X + (known to everyone) n*G where n is a tiny predictable number, then I can indeed determine which int*G tweak was used, no?

The full calculation is P = hash(shared_secret)\*G + X + n\*G. Since outside observers don't know hash(shared_secret)*G, they can't figure out n.

The recipient on the other hand *can* figure out n by calculating hash(shared_secret)\*G + X and subtracting it from P (where P is every taproot output in a tx).

Note that as soon as you start modifying shared_secret, you have to multiply by G again for every n. So scanning cost goes up linearly with n as opposed to staying the same no matter how big n is.